### PR TITLE
DATAUP-532: Add default job handlers

### PR DESCRIPTION
**Pull request by ialarmedalien**
# Description of PR purpose/changes

- Add in default job handlers to the Job Manager
- Update Job Comm Channel to take into account the new backend requests and responses
- Organise variables in JCC a bit

Two of the job status table tests have failed due to the changes from adding the default job handlers; since I will be editing that file in the next PR, I'm going to leave them as-is for now.

# Jira Ticket / Issue #
e.g. https://kbase-jira.atlassian.net/browse/DATAUP-532
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
- [x] Tests pass locally and in GitHub Actions
- [] Changes available by spinning up a local narrative - no surface changes, all behind the scenes.

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook

### CODE REVIEW

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 1]
```diff
@@ -68,6 +69,24 @@ define([
             START_UPDATE: 'start_job_update',
             STOP_UPDATE: 'stop_job_update',
         },
+        BACKEND_RESPONSES = {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:72]: Categorised/organised these message type strings a bit more.

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 2]
```diff
@@ -268,53 +277,43 @@ define([
                     break;
 
                 // CELL messages
-                case 'run_status':
-                    // Send job status notifications on the default channel,
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:272]: comment in the wrong place!

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 3]
```diff
@@ -268,53 +277,43 @@ define([
...
+
+                // errors
                 case 'job_comm_error':
                     console.error('Error from job comm:', msg);
                     if (!msgData) {
                         break;
                     }
                     jobId = msgData.job_id;
                     switch (msgData.source) {
-                        case 'cancel_job':
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:295]: use the default for this error

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 4]
```diff
@@ -268,53 +277,43 @@ define([
...
-                            });
-                            this.sendBusMessage(JOB, jobId, 'job-canceled', {
+                        case JOB_REQUESTS.LOGS:
+                            this.sendBusMessage(JOB, jobId, RESPONSES.LOGS, {
                                 jobId,
                                 error: msgData.message,
                             });
-                            break;
-                        case 'job_logs':
-                        case 'job_logs_latest':
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:306]: job logs latest no longer exists, so this is just `job_logs` now

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 5]
```diff
@@ -329,100 +328,82 @@ define([
...
-                                    jobState: jobRetried.job.state,
-                                    outputWidgetInfo: jobRetried.job.widget_info,
-                                },
-                                retry: {
-                                    jobState: jobRetried.retry.state,
-                                    outputWidgetInfo: jobRetried.retry.widget_info,
-                                },
-                            }
-                        );
+                        this.sendBusMessage(JOB, jobRetried.job.state.job_id, RESPONSES.RETRY, {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:376]: no change except to whitespace

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 6]
```diff
@@ -329,100 +328,82 @@ define([
...
+                            jobInfo: msgData[_jobId],
+                        });
                     });
                     break;
 
-                case 'job_info_batch':
-                    // TODO: awaiting backend implementation
-                    break;
-
-                case 'job_init_err':
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:360]: moved up earlier in the file

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 7]
```diff
@@ -268,53 +277,43 @@ define([
...
                                 error: msgData.message,
                             });
-                            break;
-                        case 'job_logs':
-                        case 'job_logs_latest':
                             this.sendBusMessage(JOB, jobId, 'job-log-deleted', {
                                 jobId,
                                 message: msgData.message,
                             });
-                            this.sendBusMessage(JOB, jobId, 'job-logs', {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:311]: this hasn't been deleted, I just switched the order of this message and the prior one.

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 8]
```diff
@@ -329,100 +328,82 @@ define([
...
 
                 /*
                  * The job status for one or more jobs. See job_status_all
                  * for a message which covers all active jobs.
-                 * Note that these messages are additive to the job panel
-                 * cache, but the reverse logic does not apply.
+                 *
+                 * data structure: object with key jobId and value { jobState: job.state, outputWidgetInfo: job.widget_info }
                  */
-                case 'job_status':
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:400]: there were a number of backend responses that returned job statuses; they all now use the message type `job_status`

File: kbase-extension/static/kbase/js/common/jobManager.js [hunk 9]
```diff
@@ -18,7 +16,7 @@ define(['common/jobMessages', 'common/jobs'], (JobMessages, Jobs) => {
     const jobCommand = {
         cancel: {
             command: 'request-job-cancel',
-            listener: 'job-canceled',
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobManager.js:21]: doesn't exist!

File: kbase-extension/static/kbase/js/common/jobManager.js [hunk 10]
```diff
@@ -323,6 +321,131 @@ define(['common/jobMessages', 'common/jobs'], (JobMessages, Jobs) => {
         }
     }
 
+    /**
+     * A set of generic message handlers
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobManager.js:325]: these run whenever the job manager is active; there will also be widget-specific handlers that do specific actions related to incoming messages and HTML components, e.g. the job status line updating itself or a row in the job status table reacting to a new status message.

File: kbase-extension/static/kbase/js/common/jobManager.js [hunk 11]
```diff
@@ -323,6 +321,131 @@ define(['common/jobMessages', 'common/jobs'], (JobMessages, Jobs) => {
...
+
+            handleJobDoesNotExist(self, message) {
+                const { jobId } = message;
+                self.handleJobStatus(self, {
+                    jobId,
+                    jobState: {
+                        job_id: jobId,
+                        status: 'does_not_exist',
+                    },
+                });
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobManager.js:380]: I am a little concerned about this as it seems as though there may be occasions where the FE gets an incorrect 'job-does-not-exist' message from the backend. Will investigate further once the job status table is updated.

File: kbase-extension/static/kbase/js/common/jobManager.js [hunk 12]
```diff
@@ -499,35 +622,49 @@ define(['common/jobMessages', 'common/jobs'], (JobMessages, Jobs) => {
...
-                // create the child jobs
-                const childJobs = child_job_ids.map((job_id) => {
-                    return {
-                        job_id: job_id,
-                        batch_id: batch_id,
-                        status: 'created',
-                        created: 0,
-                    };
-                });
+                const allJobIds = [batch_id].concat(child_job_ids),
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobManager.js:625]: more correct variable names

File: test/unit/spec/common/jobCommChannel-spec.js [hunk 13]
```diff
@@ -12,9 +13,15 @@ define(['common/jobCommChannel', 'base/js/namespace', 'common/runtime', 'testUti
         },
     };
     const DEFAULT_COMM = {
-        on_msg: () => {},
-        send: () => {},
-        send_shell_message: () => {},
+        on_msg: () => {
+            /* do nothing */
```
Comment (ialarmedalien) [inline test/unit/spec/common/jobCommChannel-spec.js:17]: prevent eslint / sonar cloud complaints

File: test/unit/spec/common/jobCommChannel-spec.js [hunk 14]
```diff
@@ -539,7 +635,9 @@ define(['common/jobCommChannel', 'base/js/namespace', 'common/runtime', 'testUti
                 return comm.initCommChannel().then(() => {
                     comm.handleCommMessages(msg);
                     if (test.expectedMultiple) {
-                        expect(testBus.send.calls.allArgs()).toEqual(test.expectedMultiple);
+                        expect(testBus.send.calls.allArgs()).toEqual(
+                            jasmine.arrayWithExactContents(test.expectedMultiple)
```
Comment (ialarmedalien) [inline test/unit/spec/common/jobCommChannel-spec.js:639]: handy helper to match array contents, regardless of order

File: kbase-extension/static/kbase/js/common/jobManager.js [hunk 15]
```diff
@@ -544,7 +686,13 @@ define(['common/jobMessages', 'common/jobs'], (JobMessages, Jobs) => {
...
-    class JobManager extends BatchInitMixin(jobActionsMixin(JobManagerCore)) {}
+    class JobManager extends BatchInitMixin(JobActionsMixin(DefaultHandlerMixin(JobManagerCore))) {}
 
-    return JobManager;
+    return {
+        JobManagerCore,
+        DefaultHandlerMixin,
+        JobActionsMixin,
+        BatchInitMixin,
+        JobManager,
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobManager.js:696]: export all these bits and pieces for more accurate testing

File: kbase-extension/static/kbase/js/common/jobManager.js (hunk 11 above)
Comment (briehl) [reply kbase-extension/static/kbase/js/common/jobManager.js:401]: That sounds problematic. We should know what causes that.

**Review by briehl — APPROVED** [review]
:+1:

**Comment by briehl** [comment]
Looks like a couple of job status table tests need updating after the JobManager changes.

**Comment by briehl** [comment]
And ElasticSearch seems to be wonky in the integration tests, but that's oddly not surprising.

**Pull request merged**